windows: Fix and simplify title bar padding

MatinAniss opened the pull request:
This PR fixes the off by one pixel of the top client rect when not maximized due to the added border. It also simplifies and properly fixes the title bar padding problem when maximized, it is now properly taken care of in GPUI rather then adding the padding in the UI.

Release Notes:

- N/A

ConradIrwin commented:
@JunkuiZhang Does this look good to you?

JunkuiZhang commented:
@ConradIrwin Sorry for the late response, Im in China, it was 2am midnight when you sent the message. The codes this PR modified were introduced via #9053 and #9449 , which are @zaucy 's efforts, so I know little about these.

MatinAniss committed:
fix and simplify titlebar padding

MatinAniss commented:
Fixed the merge conflict, here is some screenshots explaining the change.

Old on left, new on right.
Simplified the title bar padding issue, it is handed by GPUI properly instead of doing it in the UI. So any GPUI application does not need to do this unsafe behaviour in their UI for it to work properly. (old and new stays the same)
![zed_compare](https://github.com/zed-industries/zed/assets/76515905/f6b46095-f0be-4969-b6fa-56f964d052a6)

The other fix is the unaccounted one pixel that the top border has without a title bar. As you can see the blue border at the top of the `hello_world` example is out of the rendered client rect seen on the left. (may need to zoom in to see the top blue border)
![border_compare](https://github.com/zed-industries/zed/assets/76515905/eb6bd0a7-0955-49b9-9b4c-7c1d18dc73d4)

zaucy commented on crates/gpui/src/platform/windows/events.rs:
```diff
@@ -828,7 +835,10 @@ fn handle_hit_test_msg(
     };
     unsafe { ScreenToClient(handle, &mut cursor_point).ok().log_err() };
-    if cursor_point.y > 0 && cursor_point.y < frame_y + padding {
+    if !state_ptr.state.borrow().is_maximized()
+        && cursor_point.y > 0
+        && cursor_point.y < frame_y + padding
+    {
```
I could be wrong, but wouldn't this break the hover states of the close, max, and min buttons when you're maximized?

MatinAniss replied:
No, It prevents it from returning `HTTOP` when maximized which is "the upper-horizontal border of a window." this controls whether the resize functionality at the top of the window is enabled, which there should not be any resize (cursor/functionality) because it is maximized.

[Source](https://learn.microsoft.com/en-us/windows/win32/inputdev/wm-nchittest)

zaucy replied:
ah you're right, mb

zaucy commented:
I've stopped working on Zed for the most part, but I can take a look at this, this weekend.

Thanks for the screenshots! Does Zed work the same when in windowed mode as well with this change?

MatinAniss committed:
inclusive greater than and remove padding from resize hit

zaucy commented on crates/gpui/src/platform/windows/events.rs:
```diff
@@ -659,6 +659,13 @@ fn handle_calc_client_size(
+    if state_ptr.state.borrow().is_maximized() {
+        requested_client_rect[0].top += frame_y + padding;
+    } else {
+        // Extra pixel from the top border
```
only thing I'd maybe change is add a reason for the extra pixel

also is this 1 pixel DPI dependent? maybe mention that. Either way I'm good with it from this point on. I'll leave it to the Zed team from here on :)

MatinAniss replied:
I'm actually not sure if the thin 1 pixel border is DPI dependent, I would assume no but I don't actually know.

MatinAniss replied:
I don't know how to get the border width that works with scaling, I tried `SM_CXBORDER` but it is always reporting 1 pixel for me.

MatinAniss commented:
I made the `HTTOP` hit greater than or equal so it fixes the hit not including the first y pixel, also I removed padding as that does not seem to conform with other windows applications. I compared with Google Chrome, file explorer, terminal and it seems correct now.

MatinAniss committed:
work correctly with scaling

mikayla-maki commented:
Thank you!

shenjackyuanjie commented:
notice: this PR will break title bar rendering
like:

![image](https://github.com/zed-industries/zed/assets/54507071/4af7536c-d8fd-4b08-84a4-6cc63d262c5b)
